client.go

```diff
@@ -275,3 +292,122 @@ func (e GraphqlErrors) Error() string {
 }
+
+type parseMultipartVariablesResult struct {
+	fileMap map[string][]string
+	files   map[string]graphql.Upload
+	m       map[string]any
+}
+
+type parseMultipartVariablesStackItem struct {
+	key  string
+	path string
+	data map[string]interface{}
+}
+
+func parseMultipartVariables(variables map[string]any) parseMultipartVariablesResult {
+	stack := []parseMultipartVariablesStackItem{{key: "", data: variables}}
+
+	index := 0
+	fileMap := map[string][]string{}
+	files := map[string]graphql.Upload{}
+	for len(stack) > 0 {
+		currentItem := stack[len(stack)-1]
+		stack = stack[:len(stack)-1]
+
+		for key, value := range currentItem.data {
+			var currentPath string
+			if currentItem.path == "" {
+				currentPath = key
+			} else {
+				currentPath = currentItem.path + "." + key
+			}
+
+			switch v := value.(type) {
+			case graphql.Upload:
+				currentItem.data[key] = nil
+				fileIndex := fmt.Sprintf("file%d", index)
+				fileMap[fileIndex] = []string{fmt.Sprintf("variables.%s", currentPath)}
+				index += 1
+				files[fileIndex] = v
+			case *graphql.Upload:
+				currentItem.data[key] = nil
+				fileIndex := fmt.Sprintf("file%d", index)
+				fileMap[fileIndex] = []string{fmt.Sprintf("variables.%s", currentPath)}
+				index += 1
+				files[fileIndex] = *v
+			case map[string]any:
+				stack = append(stack, parseMultipartVariablesStackItem{key: key, data: v, path: currentPath})
+			default:
+			}
+		}
+	}
+	return parseMultipartVariablesResult{
+		fileMap: fileMap,
+		files:   files,
+		m:       variables,
+	}
+}
+
+type prepareMultipartDataResult struct {
+	buf         bytes.Buffer
+	contentType string
+}
+
+func prepareMultipartData(request *Request) (*prepareMultipartDataResult, error) {
+	res := parseMultipartVariables(request.Variables)
+
+	var fw io.Writer
+	var buf bytes.Buffer
+	mpw := multipart.NewWriter(&buf)
+	fw, err := mpw.CreateFormField("operations")
+	if err != nil {
+		return nil, err
+	}
+	input, err := json.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+	_, err = fw.Write(input)
+	if err != nil {
+		return nil, err
+	}
+	fw, err = mpw.CreateFormField("map")
+	if err != nil {
+		return nil, err
+	}
+	fileMapEntries := []string{}
+	for fileIndex, path := range res.fileMap {
+		fileMapEntries = append(fileMapEntries,
+			fmt.Sprintf(
+				"\"%s\": [\"%s\"]", fileIndex, strings.Join(path, "\",\""),
+			),
+		)
+	}
+	_, err = fw.Write([]byte(fmt.Sprintf(
+		"{%s}",
+		strings.Join(fileMapEntries, ","),
+	)))
+	if err != nil {
+		return nil, err
+	}
+	for fileIndex := range res.fileMap {
+		innerFw, fileErr := mpw.CreateFormFile(fileIndex, res.files[fileIndex].Filename)
+		if fileErr != nil {
+			return nil, fileErr
+		}
+		_, ioErr := io.Copy(innerFw, res.files[fileIndex].File)
+		if ioErr != nil {
+			return nil, ioErr
+		}
+	}
+	err = mpw.Close()
+	if err != nil {
+		return nil, err
+	}
+	contentType := mpw.FormDataContentType()
+	return &prepareMultipartDataResult{
```

Comment:
I don't think you're gaining much here with the named struct in the return, you could change the return value of the function to be the `(bytes.Buffer, string, error)` instead?